src/uu/pgrep/src/pgrep.rs

```diff
@@ -0,0 +1,288 @@
+// This file is part of the uutils procps package.
+//
+// For the full copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+
+// Pid utils
+pub mod process;
+
+use clap::{arg, crate_version, Arg, ArgAction, ArgGroup, ArgMatches, Command};
+use process::{walk_process, ProcessInformation, TerminalType};
+use regex::Regex;
+use std::{collections::HashSet, sync::OnceLock};
+use uucore::{
+    error::{UResult, USimpleError},
+    format_usage, help_about, help_usage,
+};
+
+const ABOUT: &str = help_about!("pgrep.md");
+const USAGE: &str = help_usage!("pgrep.md");
+
+static REGEX: OnceLock<Regex> = OnceLock::new();
+
+/// # Conceptual model of `pgrep`
+///
+/// At first, `pgrep` command will check the patterns is legal.
+/// In this stage, `pgrep` will construct regex if `--exact` argument was passed.
+///
+/// Then, `pgrep` will collect all *matched* pids, and filtering them.
+/// In this stage `pgrep` command will collect all the pids and its information from __/proc/__
+/// file system. At the same time, `pgrep` will construct filters from command
+/// line arguments to filter the collected pids. Note that the "-o" and "-n" flag filters works
+/// if them enabled and based on general collecting result.
+///
+/// Last, `pgrep` will construct output format from arguments, and print the processed result.
+#[uucore::main]
+pub fn uumain(args: impl uucore::Args) -> UResult<()> {
+    let matches = uu_app().try_get_matches_from(args)?;
+    let pattern = collect_arg_patterns(&matches);
+
+    // Pattern check
+    let flag_newest = matches.get_flag("newest");
+    let flag_oldest = matches.get_flag("oldest");
+
+    if (!flag_newest && !flag_oldest) && pattern.is_empty() {
+        return Err(USimpleError::new(
+            2,
+            "no matching criteria specified\nTry `pgrep --help' for more information.",
+        ));
+    }
+
+    if pattern.len() > 1 {
+        return Err(USimpleError::new(
+            2,
+            "only one pattern can be provided\nTry `pgrep --help' for more information.",
+        ));
+    }
+
+    // Verifying regex pattern
+    // And put it into static `REGEX`
+    let pattern = if matches.get_flag("exact") {
+        format!("^{}$", pattern.first().unwrap())
+    } else {
+        pattern.first().unwrap_or(&String::new()).to_string()
+    };
+    REGEX
+        .set(Regex::new(&pattern).map_err(|e| USimpleError::new(2, e.to_string()))?)
+        .unwrap();
+
+    // Collect pids
+    let pids = {
+        let mut pids = collect_matched_pids(&matches);
+        process_flag_o_n(&matches, &mut pids)
+    };
+
+    if pids.is_empty() {
+        uucore::error::set_exit_code(1);
+    }
+
+    // Processing output
+    let result = || {
```

Comment:
In Rust you can just do this:
```suggestion
    let output = {
```

Reply:
This seems to be mostly due to the newline, if you don't join with newlines, but put a newline after every entry (so at the end of each `format!`), it might become simpler.